Add ability to toggle reminders

one-new-message opened the pull request:
Obviously I can't work on implementing this in the official client, but I think this is all that is necessary to add the feature to the API. Currently drafting this PR to enable discussion, but maybe the issue tracker is a better place for this??

On Discord I shared this mockup:
![screenshot](https://github.com/ApparyllisOrg/SimplyPluralApi/assets/16503012/be169c63-9374-4fbe-b868-069b5c088b7b)

Hopefully this type of design is versatile enough (I believe it is). Though discussion of the frontend should probably be held somewhere else entirely :)

Commit:
implement ability to toggle reminders

Cellaryllis commented:
Hey @one-new-message, only saw this now.

In general this seems fine, however a few things to keep in mind:
* Fields are not a given to exist, so checking `!data.enabled` will return true on everyone who has not set that property to true (which means everyone as soon as the PR would be merged). For every property check you need to explicitly check against the default value or the inverse of the default value.

For example enabled would be true by default so we would check against the inverse, this ensures that people who don't have the property don't get their reminders turned off by default.
`date.enabled === false`

Secondly, this would still send a reminder and not delete any scheduled reminders yet-to-be-sent. You would need to delete all queued events that are scheduled for the reminders you disable.

one-new-message committed:
Merge branch 'development' of https://github.com/one-new-message/SimplyPluralApi into add-reminder-toggle

one-new-message committed:
don't require setting timer enabled/disabled on creation

one-new-message commented:
Thanks for the feedback! To start, I've removed the requirement to set the `enabled` property when creating reminders to retain compatibility with the existing client.

Changing the check to `data.enabled === false` is correct; I did not think through how `!data.enabled` would behave.

There is no need for further changes for repeated reminders--when a PATCH request is received at the `v1/timer/repeated/:id` endpoint, it calls `repeatedTimer.update()`. That async function calls `repeatRemindersEvent()` in `src/modules/events/repeatReminders.ts`, which already includes its own step of removing and rescheduling already scheduled reminders updated by the request. This rescheduling is performed by `scheduleReminder()`, which still begins with the enabled check I implemented: `if (data.enabled === false) return;`

I don't believe this affects automated reminders either. These are not queued until they are triggered, and the check I implemented prevents them from being scheduled if disabled, which means they are never added to the queue. (Unless I'm missing something with the way the routes and events are configured/controlled.)

Side note: I've managed to break my commit history in a way that may break merging again, so I'll likely need to move my changes into a new clean topic branch. If these changes are still potentially wanted, I'll draft a new PR and close this one.

one-new-message committed:
fix checks to assume reminder is enabled if `data.enabled` is absent

Cellaryllis commented:
Hey @one-new-message thanks for the update (although by now the commits have merge conflicts :see_no_evil: We did a lot of refactoring so I suspect that's what broke it. Once 1.11 is out we'll reach back out to get this PR merged and then we can update the app.

Cellaryllis commented:
Hey @one-new-message the API is now 'ready' in a sense that nothing major will change for a while so if you want to do a PR feel free.
